storops/lib/version.py

```diff
@@ -0,0 +1,305 @@
+        func_key = VersionedFuncManager.get_function_name(func)
+        versioned_func = VersionedFunc(self.criteria, func)
+        VersionedFuncManager.add_func(func_key, versioned_func)
+
+        @six.wraps(func)
+        def _inner(*args, **kwargs):
+            version = self.version_getter(Version.METHOD,
+                                          *args, **kwargs)
+
+            inner_func = VersionedFuncManager.get_support_func(
+                func_key, version)
+            return inner_func(*args, **kwargs)
+
+        return _inner
+
+    def _cls_decorator(self, cls):
+        mro_dict = [o.__dict__ for o in inspect.getmro(cls)]
+        for attr_name, attr_val in inspect.getmembers(cls):
+            if (not inspect.ismethod(attr_val) and
+                    not inspect.isfunction(attr_val)):
+                # Only handle functions
+                continue
+
+            if self._is_classmethod(attr_name, mro_dict):
+                setattr(cls, attr_name,
+                        self._handle_classmethod(attr_val))
+
+            elif self._is_staticmethod(attr_name, mro_dict):
+                setattr(cls, attr_name, self._handle_staticmethod(attr_val))
+
+            elif attr_name == '__init__':
+                setattr(cls, attr_name, self._handle_init_method(attr_val))
+
+        return cls
+
+    @staticmethod
+    def _is_classmethod(attr_name, mro_dict):
+        for mro in mro_dict:
+            if attr_name in mro and isinstance(mro[attr_name],
+                                               classmethod):
+                return True
+        return False
+
+    @staticmethod
+    def _is_staticmethod(attr_name, mro_dict):
+        for mro in mro_dict:
+            if attr_name in mro and isinstance(mro[attr_name],
+                                               staticmethod):
+                return True
+        return False
+
+    def _handle_staticmethod(self, func):
+        _inner = self._handle_method(Version.STATIC_METHOD, func)
+        return staticmethod(_inner)
+
+    def _handle_classmethod(self, func,):
+        def _inner(cls, *args, **kwargs):
+            self._test_support(Version.CLASS_METHOD,
+                               cls, *args, **kwargs)
+            return func(*args, **kwargs)
+        return classmethod(_inner)
+
+    def _handle_init_method(self, func):
+        return self._handle_method(Version.INIT_METHOD, func)
+
+    def _handle_method(self, dec_type, func):
+        def _inner(*args, **kwargs):
+            self._test_support(dec_type, *args, **kwargs)
+            return func(*args, **kwargs)
+
+        return _inner
+
+    def _test_support(self, dec_type, *args, **kwargs):
+        version = self.version_getter(dec_type, *args, **kwargs)
+        if not self.criteria.test(version):
+            raise SystemAPINotSupported(version=version)
+
+
+def get_version(dec_type, *args, **kwargs):
+    if len(args) > 0:
+        obj = args[0]
+        if (isinstance(obj, Resource) and

```

Comment:
Same here, we could just check whether `obj` has an attribute named `system_version`.  This is the only dependency we have, we don't care if it's an instance of certain class.